neps/nep-0399.md

```diff
@@ -0,0 +1,455 @@
+
+`FlatStorageManager`: owns flat storages for all shards, being stored in `NightshadeRuntime`, accepts
+ updates from `Chain` side, caused by successful processing of chunk or block.
+
+`FlatStorageCreator`: handles flat storage structs creation or initiates background creation (aka migration
+process) if flat storage data is not presend on DB yet.
+
+`FlatStateDelta`: a HashMap that contains state changes introduced in a chunk. They can be applied
+on top the state at flat storage head to compute state at another block.
+
+The reason for having separate flat storages that there are two modes of block processing, 
+normal block processing and block catchups. 
+Since they are performed on different ranges of blocks, flat storage need to be able to support
+different range of blocks on different shards. Therefore, we separate the flat storage objects
+used for different shards.
+
+### DB columns
+`DBCol::FlatState` stores a mapping from trie keys to the value corresponding to the trie keys,
+based on the state of the block at flat storage head.
+- *Rows*: trie key (`Vec<u8>`)
+- *Column type*: `ValueRef`
+
+`DBCol::FlatStateDeltas` stores all existing FSDs as mapping from `(shard_id, block_hash, trie_key)` to the `ValueRef`.
+To read the whole delta, we read all values for given key prefix. This delta stores all state changes introduced in the
+given shard of the given block.
+- *Rows*: `{ shard_id, block_hash, trie_key }`
+- *Column type*: `ValueRef`
+
+Note that `FlatStateDelta`s needed are stored in memory, so during block processing this column won't be used
+  at all. This column is only used to load deltas into memory at `FlatStorage` initialization time when node starts. 
+
+`DBCol::FlatStateMetadata` stores miscellaneous data about flat storage layout, including current flat storage
+head, current creation status and info about deltas existence. We don't specify exact format here because it is under
+discussion and can be tweaked until release.
+
+Similarly, flat head is also stored in `FlatStorage` in memory, so this column is only used to initialize
+  `FlatStorage` when node starts.
+
+### `FlatStateDelta`
+`FlatStateDelta` stores a mapping from trie keys to value refs. If the value is `None`, it means the key is deleted
+in the block.
+```rust
+pub struct FlatStateDelta(HashMap<Vec<u8>, Option<ValueRef>>);
+```
+
+```rust
+pub fn from_state_changes(changes: &[RawStateChangesWithTrieKey]) -> FlatStateDelta
+```
+Converts raw state changes to flat state delta. The raw state changes will be returned as part of the result of 
+`Runtime::apply_transactions`. They will be converted to `FlatStateDelta` to be added
+to `FlatStorage` during `Chain::postprocess_block` or `Chain::catch_up_postprocess`.
+
+### ```FlatStorageChunkView```
+```FlatStorageChunkView``` will be created for a shard `shard_id` and a block `block_hash`, and it can perform 
+key value lookup for the state of shard `shard_id` after block `block_hash` is applied.
```

Comment:
```suggestion
### `FlatStorageChunkView`
`FlatStorageChunkView` will be created for a shard `shard_id` and a block `block_hash`, and it can perform 
key value lookup for the state of shard `shard_id` after block `block_hash` is applied.
```